File: bench/mandelbrot/mandelbrot.c
/* The Great Computer Language Shootout
   http://shootout.alioth.debian.org/

   contributed by Greg Buchholz

   compile:  gcc -O2 -o mandelbrot mandelbrot.c
   run:      mandelbrot 200 > out.pbm
*/


#include<stdio.h>

int main (int argc, char **argv)
{
    int w, h, x, y, bit_num = 0;
    char byte_acc = 0;
    int i, iter = 50;
    double limit = 2.0;
    double Zr, Zi, Cr, Ci, Tr, Ti;
    
    w = atoi(argv[1]);
    h = w;

    printf("P4\n%d %d\n",w,h);

    for(y=0;y<h;y++) 
    {
        for(x=0;x<w;x++)
        {
            Zr = 0.0; Zi = 0.0;
            Cr = (2*(double)x/w - 1.5); Ci=(2*(double)y/h - 1);
        
            for (i=0;i<iter;i++)
            {
                Tr = Zr*Zr - Zi*Zi + Cr;
                Ti = 2*Zr*Zi + Ci;
                Zr = Tr; Zi = Ti;
                if (Zr*Zr+Zi*Zi > limit*limit)
                    break;
            }
        
            if(Zr*Zr+Zi*Zi > limit*limit) 
                byte_acc = (byte_acc << 1) | 0x00;
            else
                byte_acc = (byte_acc << 1) | 0x01;
                
            bit_num++; 

            if(bit_num == 8)
            {
                putc(byte_acc,stdout);
                byte_acc = 0;
                bit_num = 0;
            }
            else if(x == w-1)
            {
                byte_acc = byte_acc << (8-w%8);
                putc(byte_acc,stdout);
                byte_acc = 0;
                bit_num = 0;
            }

        }
    }	

    return(0);
}

File: bench/mandelbrot/mandelbrot.go
/* The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * contributed by The Go Authors.
 * Based on mandelbrot.c contributed by Greg Buchholz
 * flag.Arg hack by Isaac Gouy
*/

package main

import (
   "bufio"
   "flag"
   "fmt"
   "os"
   "strconv"
)

var n = 0

func main() {
   flag.Parse()
   if flag.NArg() > 0 { n,_ = strconv.Atoi( flag.Arg(0) ) }

   out := bufio.NewWriter(os.Stdout)
   defer out.Flush()

   w := n
   h := n
   bit_num := 0
   byte_acc := byte(0)
   const Iter = 50
   const Zero float64 = 0
   const Limit = 2.0

   fmt.Fprintf(out, "P4\n%d %d\n", w, h)

   for y := 0; y < h; y++ {
      for x := 0; x<w; x++ {
         Zr, Zi, Tr, Ti := Zero, Zero, Zero, Zero
         Cr := (2*float64(x)/float64(w) - 1.5)
         Ci := (2*float64(y)/float64(h) - 1.0)

         for i := 0; i < Iter && (Tr+Ti <= Limit*Limit); i++ {
            Zi = 2*Zr*Zi + Ci
            Zr = Tr - Ti + Cr
            Tr = Zr * Zr
            Ti = Zi * Zi
         }

         byte_acc <<= 1
         if Tr+Ti <= Limit*Limit {
            byte_acc |= 0x01
         }

         bit_num++

         if bit_num == 8 {
            out.WriteByte(byte_acc)
            byte_acc = 0
            bit_num = 0
         } else if x == w-1 {
            byte_acc <<= uint(8-w%8)
            out.WriteByte(byte_acc)
            byte_acc = 0
            bit_num = 0
         }
      }
   }
}
